syntax = "proto3";
option optimize_for = LITE_RUNTIME;
package proto;

message PlayerFrame {
  int64 velocity_x = 1;
  int64 velocity_y = 2;

  int64 target_x = 3;
  int64 target_y = 4;
  bool target_relative = 6;

  int32 keys = 5;
}

message Replay {
  string game_version = 1;
  int32 game_mode = 2;
  int32 players = 3;
  int32 seed = 4;

  repeated PlayerFrame player_frame = 5;
  bool can_face_secret_boss = 99;
}

message Score {
  string name = 1;
  int64 score = 2;
}

message Table {
  repeated Score score = 1;
}

message ModeTable {
  repeated Table table = 1;
}

message SaveData {
  int32 bosses_killed = 1;
  int32 hard_mode_bosses_killed = 2;

  ModeTable normal = 3;
  ModeTable hard = 4;
  ModeTable fast = 5;
  ModeTable what = 6;
  Table boss = 7;
}
